plugin/evm/header/extra.go

```diff
@@ -66,14 +157,49 @@ func VerifyExtra(rules extras.AvalancheRules, extra []byte) error {
 			)
 		}
 	default:
-		if uint64(extraLen) > params.MaximumExtraDataSize {
+		if uint64(extraLen) > MaximumExtraDataSize {
 			return fmt.Errorf(
 				"%w: expected <= %d but got %d",
 				errInvalidExtraLength,
-				params.MaximumExtraDataSize,
+				MaximumExtraDataSize,
 				extraLen,
 			)
 		}
 	}
 	return nil
 }
+
+// PredicateBytesFromExtra returns the predicate result bytes from the header's
+// extra data. If the extra data is not long enough, an empty slice is returned.
+func PredicateBytesFromExtra(rules extras.AvalancheRules, extra []byte) []byte {
+	offset := ap3.WindowSize
+	if rules.IsFortuna {
+		offset = acp176.StateSize
+	}
+
+	// Prior to Durango, the VM enforces the extra data is smaller than or equal
+	// to `offset`.
+	// After Durango, the VM pre-verifies the extra data past `offset` is valid.
+	if len(extra) <= offset {
+		return nil
+	}
+	return extra[offset:]
+}
+
+// SetPredicateBytesInExtra sets the predicate result bytes in the header's extra
+// data. If the extra data is not long enough (i.e., an incomplete header.Extra
+// as built in the miner), it is padded with zeros.
+func SetPredicateBytesInExtra(rules extras.AvalancheRules, extra []byte, predicateBytes []byte) []byte {
+	offset := ap3.WindowSize
+	if rules.IsFortuna {
+		offset = acp176.StateSize
+	}
+
+	if len(extra) < offset {
+		// pad extra with zeros
+		extra = append(extra, make([]byte, offset-len(extra))...)
+	}
+	extra = extra[:offset] // truncate extra to the offset
+	extra = append(extra, predicateBytes...)
+	return extra
```

Comment:
can we just return `append` here?

Reply:
Agreed with darioush; on the other hand, given this function is meant to return the predicate result bytes, it's confusing to update extra, hinting we are returning a new extra value. I'll change it to use `predicateResult := extra` from the beginning of the function